console_backend/src/output.rs

```diff
@@ -50,42 +50,59 @@ impl CsvLogging {
 pub type SbpLogging = cc::SbpLogging;
 
 #[derive(Debug)]
-pub enum SbpLogger {
-    Sbp(File),
-    Json(JsonEncoder<File, CompactFormatter>),
+pub struct SbpLogger<W> {
+    logger: SbpLogging,
+    out: W,
+    path: Option<PathBuf>,
 }
 
-impl SbpLogger {
-    pub fn new_sbp<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Sbp(File::create(filepath)?))
+impl<W: Write> SbpLogger<W> {
+    pub fn new(logger: SbpLogging, out: W, path: Option<PathBuf>) -> Self {
+        Self { logger, out, path }
     }
 
-    pub fn open_sbp<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Sbp(
-            OpenOptions::new().append(true).open(filepath)?,
-        ))
-    }
+    /// Log data into respective outputs, SBP or JSON format
+    ///
+    /// # Parameters:
+    /// - `frame`: The raw frame data that a message must have
+    /// - `msg`: Optional SBP depending on validity, only required in JSON format
+    ///
+    /// # Returns: the bytes serialized and whether path exists.
+    /// Error returns should be considered interruptions which breaks the callers flow.
+    pub fn serialize(&mut self, frame: &Frame, msg: Option<&Sbp>) -> Result<u16> {
+        if msg.is_none() {
+            error!("(SBP) message cannot be parsed as SBP, serializing frame instead: {frame:?}");
+        }
+        let bytes = match &self.logger {
+            SbpLogging::SBP_JSON => msg
+                .map(to_vec)
+                .and_then(|ret| ret.ok_or_log(|_| error!("error serializing SBP to JSON"))),
+            SbpLogging::SBP => Some(frame.as_bytes().to_owned()),
+        };
 
-    pub fn new_sbp_json<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Json(JsonEncoder::new(
-            File::create(filepath)?,
-            CompactFormatter,
-        )))
-    }
+        if let Some(path) = &self.path {
+            if !path.exists() {
+                return Err(anyhow!(
+                    "serializing path {} does not exist",
+                    path.display()
+                ));
+            }
+        }
 
-    pub fn open_sbp_json<P: AsRef<Path>>(filepath: P) -> Result<SbpLogger> {
-        Ok(SbpLogger::Json(JsonEncoder::new(
-            OpenOptions::new().append(true).open(filepath)?,
-            CompactFormatter,
-        )))
+        Ok(bytes
+            .map(|b| self.out.write_all(b.as_slice()).map(|_| b))
+            .transpose()
+            .ok_or_log(|e| error!("{e}"))
+            .flatten()
+            .map(|b| b.len() as u16)
+            .unwrap_or(0))
     }
+}
 
-    pub fn serialize(&mut self, frame: &Frame) -> Result<()> {
-        match self {
-            SbpLogger::Sbp(writer) => writer.write_all(frame.as_bytes())?,
-            SbpLogger::Json(logger) => logger.send(&frame.to_sbp()?)?,
-        }
-        Ok(())
+impl SbpLogging {
+    pub fn new_logger(&self, path: PathBuf) -> Result<SbpLogger<File>> {
```

Comment:
only Files for now, but `W: impl Write` 
🤔 will we ever want to write elsewhere like uploading this sbp on the fly